File: src/stages/rename.js
import { extname } from "path";
import { parseIsoDate } from "../core/datetime.js";
import { canKeep } from "../core/manifest.js";
import { logger } from "../utils/logger.js";

const monthFormatter = new Intl.DateTimeFormat("en-US", {
  month: "long",
});

/**
 * Rename stage: resolves target paths for media files based on their timestamps.
 *
 * Structure: yyyy/mm - Month/yyyymmddhhmm.ext
 * Collisions are handled with -1, -2, -3 suffixes.
 *
 * @param {Array} manifest - The manifest array from previous stages
 */
export function rename(manifest) {
  // Stage 1: filter out deletions, errors, and unsupported formats.
  const candidates = manifest.filter(
    (entry) => canKeep(entry) && entry.metadata?.status !== "skipped"
  );

  if (candidates.length === 0) {
    return;
  }

  // Stage 2: resolve earliest dates for each album
  const albumDates = resolveAlbumDates(candidates);

  // Track used paths for collision detection
  const usedPaths = new Set();

  // Stage 3: resolve target paths for each entry
  for (const entry of candidates) {
    try {
      entry.rename = resolveTargetPath(entry, usedPaths, albumDates);
    } catch (error) {
      logger.error(error.message);
      entry.rename = {
        status: "error",
        reason: error.message,
      };
    }
  }
}

// Resolves the earliest valid date for each album.
function resolveAlbumDates(candidates) {
  const albumDates = new Map();

  for (const entry of candidates.filter((e) => e.source?.type === "album")) {
    const dates = entry.metadata?.dates;
    const parsed = dates?.length > 0 ? dates.map(parseIsoDate).find(Boolean) : null;

    if (parsed) {
      const albumName = entry.source.name;
      const existing = albumDates.get(albumName);

      if (!existing || isEarlierDate(parsed, existing)) {
        albumDates.set(albumName, parsed);
      }
    }
  }

  return albumDates;
}

// Returns true if date a is earlier than date b.
function isEarlierDate(a, b) {
  if (a.year !== b.year) {
    return a.year < b.year;
  } else if (a.month !== b.month) {
    return a.month < b.month;
  } else if (a.day !== b.day) {
    return a.day < b.day;
  } else if (a.hour !== b.hour) {
    return a.hour < b.hour;
  } else if (a.minute !== b.minute) {
    return a.minute < b.minute;
  }
  return a.second < b.second;
}

// Resolves the target path for a single entry.
function resolveTargetPath(entry, usedPaths, albumDates) {
  const pad2 = (num) => String(num).padStart(2, "0");

  const dates = entry.metadata?.dates;

  // Validate that we have at least one date
  if (!dates || dates.length === 0) {
    throw new Error(`No timestamp available for ${entry.mediaPath}`);
  }

  // Try each date in order until one parses successfully
  const parsed = dates.map(parseIsoDate).find(Boolean);

  if (!parsed) {
    throw new Error(`No valid date format found for ${entry.mediaPath}`);
  }

  const { year, month, day, hour, minute, second } = parsed;
  const timestamp = `${year}${pad2(month)}${pad2(day)}-${pad2(hour)}${pad2(minute)}${pad2(second)}`;
  const ext = extname(entry.mediaPath).toLowerCase();

  // Determine directory structure based on album membership
  const isAlbum = entry.source?.type === "album";
  const albumDate = isAlbum ? albumDates.get(entry.source.name) : null;

  // Use album's earliest date for directory structure, or entry's own date
  const dirDate = albumDate ?? parsed;

  // Format: yyyy/mm - MonthName (e.g., 2023/05 - May)
  const monthName = monthFormatter.format(new Date(dirDate.year, dirDate.month - 1));
  const monthDir = `${dirDate.year}/${pad2(dirDate.month)} - ${monthName}`;

  // Build path with optional album subdirectory
  const albumDatePrefix = `${dirDate.year}${pad2(dirDate.month)}${pad2(dirDate.day)}`;
  const basePath = isAlbum
    ? `${monthDir}/${albumDatePrefix} - ${entry.source.name}/${timestamp}${ext}`
    : `${monthDir}/${timestamp}${ext}`;

  const targetPath = resolveCollision(basePath, usedPaths);

  // Track this path as used
  usedPaths.add(targetPath);

  return {
    status: "processed",
    targetPath,
  };
}

// Resolves path collisions by appending -1, -2, etc.
function resolveCollision(basePath, usedPaths) {
  if (!usedPaths.has(basePath)) {
    return basePath;
  }

  // Extract path parts for suffix insertion
  const ext = extname(basePath);
  const pathWithoutExt = basePath.slice(0, -ext.length);

  let counter = 1;
  let candidatePath;

  do {
    candidatePath = `${pathWithoutExt}-${counter}${ext}`;
    counter++;
  } while (usedPaths.has(candidatePath));

  return candidatePath;
}

File: packages/core/src/pixelkasten/stages/rename.py
"""
Rename stage — compute target paths for media files.

Ported from packages/core/src/stages/rename.js with intentional divergence:
no month directories. Format is YYYY/yyyymmdd-hhmmss.ext (loose) or
YYYY/yyyymmdd - Album Name/yyyymmdd-hhmmss.ext (album).
"""

import os
from datetime import datetime

from pixelkasten.manifest import ManifestEntry, Rename, Status


def rename(manifest: list[ManifestEntry]) -> None:
    """
    Resolve target paths for media files based on their timestamps.

    Mutates manifest entries in-place, setting entry.rename.
    Skips entries marked for deletion, error, or with skipped metadata.
    """
    # Filter to candidates: keepable and not skipped by reconcile
    candidates = [
        e
        for e in manifest
        if e.can_keep() and (e.metadata is None or e.metadata.status != Status.SKIPPED)
    ]

    if not candidates:
        return

    # Resolve earliest dates for each album
    album_dates = _resolve_album_dates(candidates)

    # Track used paths for collision detection
    used_paths: set[str] = set()

    for entry in candidates:
        try:
            entry.rename = _resolve_target_path(entry, used_paths, album_dates)
        except Exception as e:
            entry.rename = Rename(status=Status.ERROR, error=str(e))


def _resolve_album_dates(candidates: list[ManifestEntry]) -> dict[str, datetime]:
    """Resolve the earliest valid date for each album."""
    album_dates: dict[str, datetime] = {}

    for entry in candidates:
        if entry.source.type != "album":
            continue

        dates = entry.metadata.dates if entry.metadata else []
        parsed = _first_valid_date(dates)

        if not parsed:
            continue

        album_name = entry.source.name
        if not album_name:
            continue

        existing = album_dates.get(album_name)
        if not existing or parsed < existing:
            album_dates[album_name] = parsed

    return album_dates


def _first_valid_date(dates: list[str]) -> datetime | None:
    """Return the first parseable date from the list, or None."""
    for d in dates:
        try:
            return datetime.fromisoformat(d)
        except (ValueError, TypeError):
            pass
    return None


def _resolve_target_path(
    entry: ManifestEntry,
    used_paths: set[str],
    album_dates: dict[str, datetime],
) -> Rename:
    """Resolve the target path for a single manifest entry."""
    dates = entry.metadata.dates if entry.metadata else []

    if not dates:
        raise ValueError(f"No timestamp available for {entry.media_path}")

    parsed = _first_valid_date(dates)
    if not parsed:
        raise ValueError(f"No valid date format found for {entry.media_path}")

    timestamp = parsed.strftime("%Y%m%d-%H%M%S")
    ext = os.path.splitext(entry.media_path)[1].lower()

    # Determine directory structure
    album_name = entry.source.name if entry.source.type == "album" else None
    album_date = album_dates.get(album_name) if album_name else None

    # Use album's earliest date for directory, or entry's own date
    dir_date = album_date or parsed

    if album_name:
        dir_prefix = dir_date.strftime("%Y%m%d")
        base_path = f"{dir_date.year}/{dir_prefix} - {album_name}/{timestamp}{ext}"
    else:
        base_path = f"{parsed.year}/{timestamp}{ext}"

    target_path = _resolve_collision(base_path, used_paths)
    used_paths.add(target_path)

    return Rename(status=Status.PROCESSED, target_path=target_path)


def _resolve_collision(base_path: str, used_paths: set[str]) -> str:
    """Resolve path collisions by appending -1, -2, etc. before extension."""
    if base_path not in used_paths:
        return base_path

    root, ext = os.path.splitext(base_path)
    counter = 1

    while True:
        candidate = f"{root}-{counter}{ext}"
        if candidate not in used_paths:
            return candidate
        counter += 1
